src/fold/composites/optimize.py

```diff
@@ -8,64 +8,86 @@
 import pandas as pd
 from sklearn.model_selection import ParameterGrid
 
-from fold.utils.list import wrap_in_list
+from fold.traverse import traverse, traverse_apply
+from fold.utils.list import ensure_dict, to_hierachical_dict, wrap_in_list
 
-from ..base import Artifact, Optimizer, Tunable
+from ..base import Artifact, Optimizer, Pipeline, Tunable
 from .common import get_concatenated_names
 
 
 class OptimizeGridSearch(Optimizer):
-    candidates: Optional[List[Tunable]] = None
+    candidates: Optional[List[Pipeline]] = None
     selected_params_: Optional[dict] = None
-    selected_model_: Optional[Tunable] = None
+    selected_pipeline_: Optional[Tunable] = None
     scores_: Optional[List[float]] = None
     param_permutations: List[dict]
 
     def __init__(
         self,
-        model: Tunable,
-        param_grid: dict,
+        pipeline: Pipeline,
         scorer: Callable,
         is_scorer_loss: bool = True,
     ) -> None:
-        self.model = wrap_in_list(model)
-        self.param_grid = param_grid
-        self.name = "GridSearchOptimizer-" + get_concatenated_names(self.model)
+        self.pipeline = wrap_in_list(pipeline)
+        self.name = "GridSearchOptimizer-" + get_concatenated_names(self.pipeline)
         self.scorer = scorer
         self.is_scorer_loss = is_scorer_loss
-        self.param_permutations = list(ParameterGrid(self.param_grid))
 
     @classmethod
     def from_cloned_instance(
         cls,
-        model: Tunable,
-        param_grid: dict,
+        pipeline: Pipeline,
         scorer: Callable,
         is_scorer_loss: bool,
         candidates: Optional[List[Tunable]],
         selected_params_: Optional[dict],
-        selected_model_: Optional[Tunable],
+        selected_pipeline_: Optional[Tunable],
         scores_: Optional[List[float]],
     ) -> OptimizeGridSearch:
-        instance = cls(model, param_grid, scorer, is_scorer_loss)
+        instance = cls(pipeline, scorer, is_scorer_loss)
         instance.candidates = candidates
         instance.selected_params_ = selected_params_
-        instance.selected_model_ = selected_model_
+        instance.selected_pipeline_ = selected_pipeline_
         instance.scores_ = scores_
         return instance
 
-    def get_candidates(self) -> Iterable[Tunable]:
+    def get_candidates(self) -> Iterable[Pipeline]:
         if self.candidates is None:
+            all_transformation = list(traverse(self.pipeline))
+
+            for transformation in all_transformation:
+                transformation._origin_id = id(transformation)
+
+            param_grid = {
+                f"{transformation._origin_id}.{key}": value
+                for transformation in all_transformation
+                for key, value in ensure_dict(transformation.params_to_try).items()
+                if transformation.params_to_try is not None
+            }
```

Comment:
Here I'm shaping the param_grid so that we have a flat dictionary for sklearn `ParameterGrid`, this is the format:
`{model_1_id.param1: [1.0, 2.0], model_1_id.param2: 5.0, model_2_id.param1:[2.0, 3.0]}`

This is passed into `ParameterGrid` below and then converted back to a hierarchy. 

To be honest, we could shortcut this by using our own `ParameterGrid` object, but I preferred to use a tested/known solution. In all honest it is just a permutation of all and we could use `itertools` for fast implementation.